experimenter/experimenter/nimbus_ui_new/views.py

```diff
@@ -94,24 +97,61 @@ def get(self, *args, **kwargs):
         return response
 
 
+def build_experiment_context(experiment):
+    doc_base_url = "https://mozilla.github.io/metric-hub/outcomes/"
+    primary_outcome_links = [
+        (outcome, f"{doc_base_url}{experiment.application}/{outcome}")
+        for outcome in experiment.primary_outcomes
+    ]
+    secondary_outcome_links = [
+        (outcome, f"{doc_base_url}{experiment.application}/{outcome}")
+        for outcome in experiment.secondary_outcomes
+    ]
+    context = {
+        "RISK_QUESTIONS": RISK_QUESTIONS,
+        "primary_outcome_links": primary_outcome_links,
+        "secondary_outcome_links": secondary_outcome_links,
+    }
+    return context
+
+
 class NimbusExperimentDetailView(DetailView):
     model = NimbusExperiment
     template_name = "nimbus_experiments/detail.html"
     context_object_name = "experiment"
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        context["RISK_QUESTIONS"] = RISK_QUESTIONS
-
-        primary_outcomes = self.object.primary_outcomes
-        secondary_outcomes = self.object.secondary_outcomes
-        doc_base_url = "https://mozilla.github.io/metric-hub/outcomes/"
-        context["primary_outcome_links"] = [
-            (outcome, f"{doc_base_url}{self.object.application}/{outcome}")
-            for outcome in primary_outcomes
-        ]
-        context["secondary_outcome_links"] = [
-            (outcome, f"{doc_base_url}{self.object.application}/{outcome}")
-            for outcome in secondary_outcomes
-        ]
+        experiment_context = build_experiment_context(self.object)
+        context.update(experiment_context)
+        context["qa_edit_mode"] = self.request.GET.get("edit_qa_status") == "true"
+        if context["qa_edit_mode"]:
+            context["form"] = QAStatusForm(instance=self.object)
         return context
+
+
+class QAStatusUpdateView(FormView):  # type: ignore[type-arg]
+    form_class = QAStatusForm
+    template_name = "nimbus_experiments/detail.html"
+
+    def form_valid(self, form):
+        form.save()
+        return super().form_valid(form)
+
+    def form_invalid(self, form):
+        context = self.get_context_data(form=form)
+        experiment = NimbusExperiment.objects.get(slug=self.kwargs["slug"])
+        context["experiment"] = experiment
+        experiment_context = build_experiment_context(experiment)
+        context.update(experiment_context)
+        context["qa_edit_mode"] = True
+        return self.render_to_response(context)
+
+    def get_success_url(self):
+        return reverse("nimbus-new-detail", kwargs={"slug": self.kwargs["slug"]})
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        experiment = NimbusExperiment.objects.get(slug=self.kwargs["slug"])
+        kwargs["instance"] = experiment
+        return kwargs
```

Comment:
This is what [UpdateView](https://docs.djangoproject.com/en/5.0/ref/class-based-views/generic-editing/#django.views.generic.edit.UpdateView) is for.  You should be able to just subclass that, override a couple class properties and that's it.  Everything you're doing here is built in.

Reply:
my bad